File: code/src/main/java/it/unifi/nave/data/Block.java
package it.unifi.nave.data;

import java.util.ArrayList;
import java.util.Collections;
import java.util.List;

public class Block {
  private BlockHeader blockHeader;
  private int eventsNum;
  private List<EventContainer> eventContainers;

  public Block(String previousHash, int difficulty) {
    blockHeader = new BlockHeader(previousHash, difficulty);
    eventContainers = new ArrayList<>();
  }

  public void addEvent(EventContainer eventContainer) {
    addEvents(Collections.singletonList(eventContainer));
  }

  public void addEvents(List<EventContainer> eventContainers) {
    this.eventContainers.addAll(eventContainers);
    eventsNum = eventContainers.size();
    updateRootHash();
  }

  private void updateRootHash() {
    MerkleTree merkleTree = new MerkleTree(eventContainers);
    blockHeader.setRootHash(merkleTree.getRootHash());
  }

  public void mine() {
    while (!blockHeader.isMined()) {
      blockHeader.incrementNonce();
    }
  }
}

File: UniBlock/src/main/scala/data/Block.scala
package it.unifi.nave.uniblock
package data

class Block(previousHash: String, difficulty: Int) {
  private val _blockHeader = new BlockHeader(previousHash, difficulty)
  private var _eventContainers: List[EventContainer] = List.empty

  def blockHeader: BlockHeader = _blockHeader

  def eventContainers: List[EventContainer] = _eventContainers

  def addEvent(eventContainer: EventContainer): Unit = {
    addEvents(eventContainer :: Nil)
  }

  def addEvents(eventContainers: List[EventContainer]): Unit = {
    _eventContainers ++= eventContainers
    _blockHeader.rootHash = MerkleTree.rootHash(_eventContainers)
  }

  def mine(): Unit = while (!_blockHeader.isMined) _blockHeader.incrementNonce()

  override def toString = s"Block($blockHeader, $eventContainers)"
}
